File: CTFlearn/Bite-code/solve.py
#!/usr/bin/python3
# -*- coding=utf8 -*-
"""
# @Author : iceylemon
# @Created Time : 2022-04-07 16:12:53
# @Description : 
"""

import numpy as np

# Simple tutorial to Java Bytecode https://dzone.com/articles/introduction-to-java-bytecode
# iload_id: load local variable id(an integer) to stack
# istore_id: pop the top of stack and make it local variable id

def checkNum(iload_0):
    istore_1 = np.int32(iload_0 << 3)
    istore_2 = np.int32(iload_0 ^ 525024598)
    result = np.int32(istore_1 ^ istore_2)
    if result == -889275714:
        print("flag is", iload_0)

tmp = np.uint32(-889275714)
tmp ^= np.uint32(525024598)
bin_data = list(bin(tmp)[2:])
flag = bin_data[-3:][::-1]

for i in range(len(bin_data) - 4, -1, -1):
    bit = bin_data[i]
    bit = int(bit) ^ int(bin_data[i + 3])
    flag += [str(bit)]
    bin_data[i] = str(bit)

flag = flag[::-1]
print('flag is', np.int32(int(''.join(flag), 2)))

File: CTFlearn/Bite-code/solve.cpp
#include "iostream"
#include "bitset"
#include "climits"
using namespace std;

int checkNum(int xt) {
    unsigned int x = xt;
    unsigned int a = x << 3;
    unsigned int b = x ^ 525024598;
    unsigned int result = a ^ b;
    unsigned int tmp = -889275714;
    if(result == tmp) return xt;
    return 0;
}

signed main() {
    for(int i = INT_MIN; i <= INT_MAX; i ++) {
        int a = checkNum(i);
        if(a) {
            cout << "flag is " << a << endl;
            return 0;
        }
    }
}